### PR TITLE
Fixes shellcheck errors/warnings and some legit bugs found in ./packaging/installer/install-required-packages.sh

### LINKED ISSUE #6967: Kickstart.sh doesn't work for CentOS 8 & RHEL 8
##### Bug report summary
Running the `kickstart.sh` on CentOS 8 fails on 'Installation of required packages'.

```
/etc/os-release information:
NAME            : CentOS Linux
VERSION         : 8 (Core)
ID              : centos
ID_LIKE         : rhel fedora
VERSION_ID      : 8

We detected these:
Distribution    : centos
Version         : 8
Codename        : 8 (Core)
Package Manager : install_yum
Packages Tree   : centos
Detection Method: /etc/os-release
Default Python v: 2

This system is receiving updates from Red Hat Satellite or Spacewalk server.
Last metadata expiration check: 2:42:06 ago on Fri Sep 27 14:58:16 2019.
No match for argument: Judy-devel
No match for argument: MySQL-python
No match for argument: autoconf-archive
No match for argument: autogen
No match for argument: libmnl-devel
No match for argument: libuv-devel
No match for argument: python
There are following alternatives for "python": python2, python36
No match for argument: python-psycopg2



We are very sorry!

Installation of required packages failed.

What to do now:

1. Make sure your system is updated.
Most of the times, updating your system will resolve the issue.


2. If the error message is about a specific package, try removing
that package from the command and run it again.
Depending on the broken package, you may be able to continue.


3. Let us know. We may be able to help.
Open a github issue with the above log, at:


https://github.com/firehol/netdata/issues                                                                                             
```
##### OS / Environment
CentOS 8
##### Netdata version (ouput of `netdata -V`)
1.10.0 (I am limited to this version due to company policies. Version should be irrelevant to this issue) 
##### Component Name
Installation scripts
##### Steps To Reproduce
`bash kickstart.sh all --non-interactive`
##### Expected behavior
Installation should succeed.

### PR DESCRIPTION
**Pull request by prologic**
Partially Fixes #6967

**Depends on**: #7915 (_review that one first_)

##### Summary

As per title

##### Component Name

area/ci
area/packaging

##### Additional Information

2/3 PR(s) inbound

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

File: packaging/installer/install-required-packages.sh [hunk 1]
```diff
@@ -129,1282 +129,1223 @@ EOF
...
+    echo >&2 "Running in non-interactive mode"
+    # http://serverfault.com/questions/227190/how-do-i-ask-apt-get-to-skip-any-interactive-post-install-configuration-steps
+    export DEBIAN_FRONTEND="noninteractive"
+    opts="${opts} -yq"
+  fi
+
+  # install the required packages
+  for pkg in "${@}"; do
+    [[ ${DRYRUN} -eq 0 ]] && echo >&2 "Adding package ${pkg}"
+    run ${sudo} apt-get "${opts}" install "${pkg}"
```
Comment (knatsakis) [inline packaging/installer/install-required-packages.sh]: I think this should be:
```suggestion
    run ${sudo} apt-get ${opts} install "${pkg}"
```

File: packaging/installer/install-required-packages.sh [hunk 2]
```diff
@@ -129,1282 +129,1223 @@ EOF
...
+    # man dnf
+    opts="-y"
+  fi
+
+  # install the required packages
+  # --setopt=strict=0 allows dnf to proceed
+  # installing whatever is available
+  # even if a package is not found
+  opts="$opts --setopt=strict=0"
+  run ${sudo} dnf "${opts}" install "${@}"
```
Comment (knatsakis) [inline packaging/installer/install-required-packages.sh]:
```suggestion
  run ${sudo} dnf ${opts} install "${@}"
```

File: packaging/installer/install-required-packages.sh [hunk 3]
```diff
@@ -129,1282 +129,1223 @@ EOF
...
+  local opts="--ask"
+  if [ "${NON_INTERACTIVE}" -eq 1 ]; then
+    echo >&2 "Running in non-interactive mode"
+    opts=""
+  fi
+
+  # install the required packages
+  for pkg in "${@}"; do
+    [[ ${DRYRUN} -eq 0 ]] && echo >&2 "Adding package ${pkg}"
+    run ${sudo} emerge "${opts}" -v --noreplace "${pkg}"
```
Comment (knatsakis) [inline packaging/installer/install-required-packages.sh]:
```suggestion
    run ${sudo} emerge ${opts} -v --noreplace "${pkg}"
```

File: packaging/installer/install-required-packages.sh [hunk 4]
```diff
@@ -129,1282 +129,1223 @@ EOF
...
+
+  local opts="--force-broken-world"
+  if [ "${NON_INTERACTIVE}" -eq 1 ]; then
+    echo >&2 "Running in non-interactive mode"
+  else
+    opts="${opts} -i"
+  fi
+
+  # install the required packages
+  run ${sudo} apk add "${opts}" "${@}"
```
Comment (knatsakis) [inline packaging/installer/install-required-packages.sh]:
```suggestion
  run ${sudo} apk add ${opts} "${@}"
```

File: packaging/installer/install-required-packages.sh [hunk 5]
```diff
@@ -1413,100 +1354,97 @@ install_equo() {
...
+
+  local opts="--ignore-unknown"
+  if [ "${NON_INTERACTIVE}" -eq 1 ]; then
+    echo >&2 "Running in non-interactive mode"
+    # http://unix.stackexchange.com/questions/82016/how-to-use-zypper-in-bash-scripts-for-someone-coming-from-apt-get
+    opts="${opts} --non-interactive"
+  fi
+
+  # install the required packages
+  run ${sudo} zypper "${opts}" install "${@}"
```
Comment (knatsakis) [inline packaging/installer/install-required-packages.sh]:
```suggestion
  run ${sudo} zypper ${opts} install "${@}"
```

File: packaging/installer/install-required-packages.sh [hunk 6]
```diff
@@ -129,1282 +129,1223 @@ EOF
...
+  ['alpine']="py-requests"
+  ['arch']="python2-requests"
+  ['centos']="python-requests"
+  ['debian']="python-requests"
+  ['gentoo']="dev-python/requests"
+  ['sabayon']="dev-python/requests"
+  ['rhel']="python-requests"
+  ['suse']="python-requests"
+  ['default']="python-requests"
+  [\
```
[the remark this answers is not in the document]
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Stacked PR; Fixed it in the bottom pR :)

File: packaging/installer/install-required-packages.sh (hunk 1 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Ahh actually according to [SC2086](https://github.com/koalaman/shellcheck/wiki/SC2086) we should use an array expansion here `${foo[@]}` at least in Bash.

File: packaging/installer/install-required-packages.sh (hunk 2 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Same here

File: packaging/installer/install-required-packages.sh (hunk 3 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Same

File: packaging/installer/install-required-packages.sh (hunk 4 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Ditto

File: packaging/installer/install-required-packages.sh (hunk 5 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Ditto

**Comment by prologic** [comment]
Re-testing this PR

**Comment by prologic** [comment]
> Re-testing this PR

Works for CentOS, Debian and Alpine. Changes are similar for other systems.

File: packaging/installer/install-required-packages.sh [hunk 7]
```diff
@@ -129,1282 +129,1222 @@ EOF
...
+  [ -n "${pacman}" ] && echo >&2 " - Arch Linux based (installer is: pacman)" && opts="pacman ${opts}"
+  [ -n "${emerge}" ] && echo >&2 " - Gentoo based (installer is: emerge)" && opts="emerge ${opts}"
+  [ -n "${equo}" ] && echo >&2 " - Sabayon based (installer is: equo)" && opts="equo ${opts}"
+  [ -n "${apk}" ] && echo >&2 " - Alpine Linux based (installer is: apk)" && opts="apk ${opts}"
+  echo >&2
+
+  REPLY=
+  while [ -z "${REPLY}" ]; do
+    echo "To proceed please write one of these:"
+    echo "${opts// /,/}"
```
[the remark this answers is not in the document]
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: Okay :) Committed locally 😀

**Comment by prologic** [comment]
Opted for the proper fix (_sorry about the race condition_) of:

```sh
local opts=
opts="$opts ..."
read -a -r foo_opts <<< "$opts"
run ... "${foo_opts[@]}"
```

Tested on:

- Debian 9
- Debian 10
- CentOS 6
- CentOS 7
- Alpine 3.8
- Alpine 3.9

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

File: packaging/installer/install-required-packages.sh [hunk 8]
```diff
@@ -33,19 +32,20 @@ PACKAGES_NETDATA_SENSORS=${PACKAGES_NETDATA_SENSORS-0}
...
-yum=$(which yum 2>/dev/null || command -v yum 2>/dev/null)
-zypper=$(which zypper 2>/dev/null || command -v zypper 2>/dev/null)
+apk=$(command -v apk 2> /dev/null)
+apt_get=$(command -v apt-get 2> /dev/null)
+dnf=$(command -v dnf 2> /dev/null)
+emerge=$(command -v emerge 2> /dev/null)
+equo=$(command -v equo 2> /dev/null)
+pacman=$(command -v pacman 2> /dev/null)
+yum=$(command -v yum 2> /dev/null)
+zypper=$(command -v zypper 2> /dev/null)
```
Comment (ilyam8) [inline packaging/installer/install-required-packages.sh]: is it really a good thing to have a space after `>` ?

i think `2>/dev/null` is more common then `2> /dev/null`

Comment (ilyam8) [reply packaging/installer/install-required-packages.sh]: (same for all redirects), imo space is not needed

**Comment by ilyam8** [comment]
> local opts=
opts="$opts ..."
read -a -r foo_opts <<< "$opts"
run ... "${foo_opts[@]}"

can be simplified to (not sure if it is bash specific thing)

```bash
opts=("$@")
```

File: packaging/installer/install-required-packages.sh (hunk 8 above)
Comment (prologic) [reply packaging/installer/install-required-packages.sh]: `shfmt` does this. We're pretty much letting tools take care of formatting / code-style for us with agreed upon options. @Ferroin is doing to put up a PR in our docs repo for this.

**Comment by prologic** [comment]
> > local opts=
> > opts="$opts ..."
> > read -a -r foo_opts <<< "$opts"
> > run ... "${foo_opts[@]}"
> 
> ~can be simplified to (not sure if it is bash specific thing)~
> 
> ```shell
> opts=($opts)
> ```
> 
> ok, not good, this is https://github.com/koalaman/shellcheck/wiki/SC2206

Sorry I'm missing some context here. Did you find a bug with this PR?

**Comment by ilyam8** [comment]
No, i said that `opts=($opts)` does same job and looks easier then `read -a -r foo_opts <<< "$opts"` but then realized that i am wrong and it is bad things (after reading bash best practices)

**Comment by prologic** [comment]
> No, i said that `opts=($opts)` does same job and looks easier then `read -a -r foo_opts <<< "$opts"` but then realized that i am wrong and it is bad thing (after reading bash best practices).
> 
> Felt myself ready to give suggestion after a weekend with bash basically 😄 🙈

Ahh  ! 😃 Don't worry! 🙌

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4912472&bid=16261729)

**Comment by ilyam8** [comment]
@prologic it is 

```diff
+0 −0 
```

> There are no new changes to show.

**Comment by ilyam8** [comment]
@Ferroin approved 😄

**Pull request closed without merging**